### PR TITLE
SPESS PARALLAX - "Fuck you git" edition

### PR DESCRIPTION
**Pull request by monster860**
[https://embed.gyazo.com/abfba09d19ea42651a95c147b715ccfd.gif](https://embed.gyazo.com/abfba09d19ea42651a95c147b715ccfd.gif)

510 branch is merged, and git is fucking with me, so reopening time.

Profiler report:
http://pastebin.com/WuwMLFiD
Server CPU usage of parallax is minimal. If it causes client lag, it can be turned off in prefs, which removes the BLEND_MULTIPLY layer, and the client images.

:cl: monster860
rscadd: Adds a parallax effect to space, which can be turned off in prefs
/:cl:

### CODE REVIEW

**Comment by Spacemanspark** [comment]
:+1:

**Comment by Regen1** [comment]
Could you give us a profiler report of this?

It kinda looks like it's stuttering..

**Comment by Aurorablade** [comment]
Even with a config option...I am still not sure of this.

**Comment by monster860** [comment]
@Regen1 This is a gif @FlattestGuitar made ages ago from an older version of the PR. The key word here is "gif"

The performance impact should be minimal, since I've reduced the number of /area/space instances down to 1, and all the /turf/space tiles that are not in /area/space are pretty much shuttles (They have spaceturfs under the engines) and solars.

**Comment by Regen1** [comment]
>should be minimal

Then by all means, show that

**Comment by monster860** [comment]
Doing so.... *waits for enormous map to load.*

**Comment by monster860** [comment]
http://pastebin.com/WuwMLFiD

**Comment by tigercat2000** [comment]
I still don't like this, personally, and let me show you why:
This is what it looks like when you are just going around near the station, and it's great n' all:
https://puu.sh/o4FXj/cc74b4b59d.webm

But as soon as you end up in space, this is the effect produced:
https://puu.sh/o4FMx/f8cec92700.webm
It looks like the video is stuttering, but pay attention to the heartbeat meter on the right. That's a really nasty-looking effect, if you ask me.

It's much *better* now that you can turn it off, but I still question if it's really good to merge in the first place.

**Comment by monster860** [comment]
Is it possible to animate() screen_loc?

**Comment by tigercat2000** [comment]
No. I wish you could, that would make a lot of things easier.

**Comment by monster860** [comment]
Hm.

Well, BYOND is unable to comprehend any units smaller than a pixel, so it's probably gonna stay that choppy way.

**Comment by Regen1** [comment]
I still like this change
:+1: from me

I like the aesthetic feeling it gives, not being able to navigate yourself in space would be a fun addition and be more realistic

**Comment by KasparoVy** [comment]
It's a preference
If you're drifting in deep spess and you want to see where you're going just turn it off I guess lol

Still wicked neat.

**Comment by monster860** [comment]
You can still tell which way you're going though. The layer moves 1 pixel for every 1 tile you move.

**Comment by Fox-McCloud** [comment]
Still a :-1: from me---even more so now that @tigercat2000 showed what it looks like when you're in full space.

One thing also worth mentioning is that this kills of space animations themselves--no more twinkling starlight or dynamicness to it; just a solid background that never  changes.

**Comment by Spacemanspark** [comment]
I'd rather have this than the twinkling background.

**Comment by monster860** [comment]
@Fox-McCloud There's an option in the prefs for this. If people have a problem with the slight choppiness, they can turn it off.

**Comment by Necaladun** [comment]
If it's an option in preferences, great. Let the people choose!

However Fox mentioned performance issues. Is that client side that'd change with prefs, or server side?

**Comment by monster860** [comment]
Turning it on puts about ~42, depending on how many shuttles and solars there are, /image objects to the player.

Having 100 players with 42 image objects isn't that bad, considering how many image objects the AI has (~4096) that are constantly being removed and readded.

**Comment by PKPenguin321** [comment]
this is the coolest fucking thing ever i wish we could port this at /tg/
god damn you guys this is radical
like, damn

**Comment by Earthdivine** [comment]
If its a prefrence toggle, then it can easily be improved on to add the whole twinkling lights thing at a later date. 

@monster860 42x100 is still 4200 images, which is still kinda bad.

**Comment by VampyrBytes** [comment]
As a preference, it's :+1: from me

**Comment by sasanek12** [comment]
:+1: from me, as preference.

**Comment by Necaladun** [comment]
As a preference, there's no argument really about how it looks, as some people clearly like it.

It's only a performance question now. If it's correct that it's 100th of what an AI has, I see no problem there.

**Comment by monster860** [comment]
I just realised all players get the same 42 images. So its only 42.

**Comment by pinatacolada** [comment]
:+1: even when in deep space, it reminds me of flying over water at high altitudes, nothing seems to move much. If you really need to navigate in space, you'd be better off with a GPS anyway

**Comment by Dumbdumn5** [comment]
Absofuckin'lutely, stuttering in deep space or not, this is a :+1: :x: :100: from me.

**Comment by Fox-McCloud** [comment]
Are space tiles even randomly selected anymore or is it all preset.

Also, I'm not going to support this if twinkling space/starlight is removed.

**Comment by Spacemanspark** [comment]
Twinkling is a small little detail that we won't really even notice.
I don't see why it would override something as awesome as this.

**Comment by Aurorablade** [comment]
No. Gets rid of twinkling. Is called in a mobs move proc leaving mobs like pais in bags and borers out among other things. Is pretty much just useing several layers at one time. As a concept it's fine but the code execution no just no

**Comment by CrAzYPiLoT-SS13** [comment]
IMO, this is great. I see no problems with the code, especially if it only
uses 42 images. Also, Fox, I didn't even know we had twinkling, and I've
been playing at Paradise prolly over a year now. :+1:

File: code/_onclick/hud/hud.dm [hunk 1]
```diff
@@ -127,6 +127,62 @@ datum/hud/New(mob/owner)
...
+
+var/area/global_space_area = null
+
+/obj/screen/spessbg
+	var/offset_x = 0
+	var/offset_y = 0
+/datum/hud/var/showing_parallax
+/datum/hud/proc/create_parallax()
+	var/client/C = mymob.client
+	showing_parallax = C.prefs.space_parallax
```
Comment (tigercat2000) [inline code/_onclick/hud/hud.dm]: this does literally nothing

**Comment by tigercat2000** [comment]
:-1:
 - This removes all animations from space turfs.
 - Deep-space travel looks like the entire server is choppy.
 - Hooked into atom/Move() and mob/Move(), but doesn't even work for things like pAI's, MMI's, even when they are in the user's pocket.
 - Directly adds all space turfs to a "global space area" (100.5% sure that adding turfs directly to an area contents, especially when they are also in a different area results in undefined behaviour)
 - Every client redefines the `spessbg` objects variables instead of using New() and having the literally hardcoded variables only defined on the object.
 - Hides the animated objects for clients with parallax with plane `-100` instead of 
  - not creating them
  - keeping them off the client's screen.

**Comment by monster860** [comment]
Fixed the pAI's, MMI's.

Adding turfs directly to an areas contents is the correct and only way to set turfs to an area, and is used in shuttles, station blueprints, and other shit.

**Comment by monster860** [comment]
@Fox-McCloud @Aurorablade If people really have that much of a problem with lack of twinkling, they can turn it off.

**Comment by tigercat2000** [comment]
@monster860 The difference is, shuttles only override a small area. This adds every space turf to the same area instance directly, and given that it's the `contents`, that means that every space turf has a `loc` of the global_space_area. Areas are definitely *not* meant to work cross-z-level.

It seems I wasn't paying attention in my testing, and space turfs with this turned off *do* have animated icons @Fox-McCloud @Aurorablade.

File: code/game/atoms_movable.dm [hunk 2]
```diff
@@ -129,6 +133,12 @@
 
 	for(var/datum/light_source/L in light_sources)
 		L.source_atom.update_light()
+	
+	for(var/mob/M in GetAllContents()) // Update parallax of contents
+		if(M.hud_used)
+			M.hud_used.update_parallax()
+	if(ismob(src) && src:hud_used)
```
Comment (tigercat2000) [inline code/game/atoms_movable.dm]: Define mob/forceMove and call the parent, never override type saftey with colon operators.

**Comment by monster860** [comment]
As far as I can tell, giving every space turf the same area works fine. Nothing seems broken on the other Z-levels, I've gone to z-5's asteroid multiple times, nothing's broken.

**Comment by Tastyfish** [comment]
Changing the preference mid-game doesn't change the state of the overlays.

**Comment by Tastyfish** [comment]
For some reason, the shuttle tile to the right of the front window is rendering weird.

Preview

**Comment by monster860** [comment]
;Looks like a lighting problem to me

File: code/modules/client/preferences.dm [hunk 3]
```diff
@@ -1576,6 +1578,9 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 					var/UI_style_alpha_new = input(user, "Select a new alpha(transparence) parameter for UI, between 50 and 255", UI_style_alpha) as num
 					if(!UI_style_alpha_new | !(UI_style_alpha_new <= 255 && UI_style_alpha_new >= 50)) return
 					UI_style_alpha = UI_style_alpha_new
+				if("parallax")
+					space_parallax = !space_parallax
+					user << "Space Parallax is now [space_parallax ? "on" : "off"]."
```
Comment (tigercat2000) [inline code/modules/client/preferences.dm]: to_chat(user, "Space Parallax is now [space_parallax ? "on" : "off"].")

**Comment by Kevak** [comment]
Losing the twinkling matters quite a lot, is best animation. 

Also making it functionally impossible to navigate space without a GPS is a huge nono in my books. It would be a maaassive pain in the ass for mechanics, pod pilots, and pretty much anyone who goes EVAing. Currently it _Is_ possible to navigate without a gps. Making something hard even harder is downright silly.

Have it default to off while you sort out the various code bugs and re-institute the sparkle an I'd approve of it. I wouldn't use it however. It should default to off.

**Comment by Spacemanspark** [comment]
If you turn off the parallax, it turns on the sparkle as well.

**Comment by monster860** [comment]
@Kevak The parallax layer moves 1 pixel for every tile you move, so it's still possible to navigate space. And since the texture is the same every time, you can use it as a sort of GPS.

If you *really* want it that badly, I could make an animated space background, but it would take a bit of work to do so.

**Comment by CrAzYPiLoT-SS13** [comment]
@monster860 I could make one, if you want me to, just need to tell me how
the logic behind /image works as I haven't fiddled with it before.

**Comment by monster860** [comment]
I fixed everything that has been thrown at me that doesn't involve "no twinkle i ded"

**Comment by monster860** [comment]
At this point, I believe it is ready for merge. Admins like it, and the ones that don't can turn it off. @DZD @Fox-McCloud

**Comment by Spacemanspark** [comment]
Please merge this.

File: code/_onclick/_defines.dm [hunk 4]
```diff
@@ -1 +1,7 @@
-#define CLICKCATCHER_PLANE -99
\ No newline at end of file
+#define HIDDEN_PLANE -100
```
Comment (tigercat2000) [inline code/_onclick/_defines.dm]: this is no longer used

File: code/modules/client/preferences.dm [hunk 5]
```diff
@@ -1576,6 +1578,9 @@ var/global/list/special_role_times = list( //minimum age (in days) for accounts
 					var/UI_style_alpha_new = input(user, "Select a new alpha(transparence) parameter for UI, between 50 and 255", UI_style_alpha) as num
 					if(!UI_style_alpha_new | !(UI_style_alpha_new <= 255 && UI_style_alpha_new >= 50)) return
 					UI_style_alpha = UI_style_alpha_new
+				if("parallax")
```
Comment (tigercat2000) [inline code/modules/client/preferences.dm]: space above this

File: code/game/turfs/space/space.dm [hunk 6]
```diff
@@ -15,7 +15,21 @@
 
 /turf/space/New()
 	. = ..()
-
+	if(!global_space_area)
+		global_space_area = new /area/space()
+		global_space_area.white_overlay = image(loc=global_space_area, icon='icons/turf/space.dmi', icon_state="white")
+		global_space_area.white_overlay.plane = SPACE_WHITE_PLANE
+	if(istype(loc,/area/space))
```
Comment (tigercat2000) [inline code/game/turfs/space/space.dm]: space above this

File: code/game/turfs/space/space.dm [hunk 7]
```diff
@@ -15,7 +15,21 @@
 
 /turf/space/New()
 	. = ..()
-
+	if(!global_space_area)
+		global_space_area = new /area/space()
+		global_space_area.white_overlay = image(loc=global_space_area, icon='icons/turf/space.dmi', icon_state="white")
```
Comment (tigercat2000) [inline code/game/turfs/space/space.dm]: spacing with args, `loc = global_space_area`

**Comment by monster860** [comment]
> just a comment that says "okay whatever please merge this"

Can I have that comment now please?

**Comment by tigercat2000** [comment]
@Fox-McCloud myeh this is mildly better now and the planes n' shit seem okay for their purpose so this is okay i guess :neutral_face:

**Comment by monster860** [comment]
One more thing, I have a quick corner-shuttle turf / catwalk fix

**Comment by Aurorablade** [comment]
No. My opinion here. Not until some sort of performance test is done.

**Comment by Spacemanspark** [comment]
If it's client side only and doesn't you can toggle it on/ off, I see no reason not to merge it.

**Comment by tigercat2000** [comment]
@Spacemanspark it has potential performance concerns affecting the entire server, not just clients that turn it on, primarily in atom movement

**Comment by Spacemanspark** [comment]
Ah, fair enough.
...Test when.

**Comment by monster860** [comment]
Here's your test right here. I filled the entire admin testing area with foxes, which is pretty big. Here's the profiler report

http://pastebin.com/Xe0Bt3kA

As you can see, get_mobs_in_contents is not much impact.

**Comment by monster860** [comment]
Alright, I actually have an idea on how to optimize this.

**Comment by monster860** [comment]
Arrright, this right here is the fastest it will get and still have it reliably shift the parallax layer. @Fox-McCloud @Aurorablade @TheDZD 

This should address any potential performance concerns here.

**Comment by tigercat2000** [comment]
you've lost my approval adding Life() code and trying to make a ghetto contents list just to avoid the movement updating code

**Comment by monster860** [comment]
Its faster and just as reliable. If you have a better place to hook thsn `Life()`, please tell me

**Comment by monster860** [comment]
I request review from @Fox-McCloud

**Comment by monster860** [comment]
Since @tigercat2000 really insists on recursive searches running every tick... Here it is!

**Comment by monster860** [comment]
@Fox-McCloud @TheDZD @tigercat2000 Is there anything else needing to be done, or what?

File: code/_onclick/hud/hud.dm [hunk 8]
```diff
@@ -149,26 +148,31 @@ var/area/global_space_area = null
 	layer = AREA_LAYER
 	plane = SPACE_LAYER_PLANE
 
+/obj/screen/pmaster_whitespace
+	plane = SPACE_TURF_PLANE
+	color = list(0, 0, 0, \
```
Comment (tigercat2000) [inline code/_onclick/hud/hud.dm]: minus the `\`
invocations like list() are inherently multiline

**Comment by TheDZD** [comment]
Oh god, merge conflicts again.

**Comment by TheDZD** [comment]
Will try a test merge of this at some point while Necaladun is around to patch, will run for a round to profile it. If it causes huge performance problems, or causes lots of client crashes, will revert and re-patch.

**Comment by monster860** [comment]
Hm, shit's being broken

**Comment by monster860** [comment]
Morgue conflicts fixed.

**Comment by Fox-McCloud** [comment]
and what happens when you have an expanded view range greater than 7x7

**Comment by monster860** [comment]
@Fox-McCloud Things break. But a lot of things break such as:

- Blindness
- Being blinded by a flash
- Welding gear
- The HUD will now be slightly off
- Nuke cinematic
- Taking damage
- Going into crit
- The ability to have the whole thing be in `64x64` mode, and fit perfectly when you use a 1080p monitor
- The shuttle transit turfs don't extend far enough. You can see that the transit turfs are a lie! ( Although this can be fixed with the allocatable shuttle turfs system)
- Nuke ops "fourth wall" is now visible.
- The title screen needs changes

So yeah, if you have a view greater than 15x15, then a lot of things break. However, as the 511-stuff on the forum has taught us, having a couple of 480x480 layers is better than a fuckton (900) small tiles, because changing appearances is expensive.

**Comment by tigercat2000** [comment]
@monster860 all of those are able to be ignored
glaring white shit everywhere isn't

**Comment by monster860** [comment]
@tigercat2000 Then move the layer over a couple of pixels.

**Comment by Iamgoofball** [comment]
reccomend https://github.com/d3athrow/vgstation13/pull/10085 instead

**Comment by DeityLink** [comment]
You should really wait until we've merged it and confirmed that it doesn't cause aopcalyptic lag on high pop before trying to port my "port".

**Comment by monster860** [comment]
I actually want to make a planet layer now.

**Comment by monster860** [comment]
Update: We've got planet!

**Comment by tigercat2000** [comment]
I'd rather wait for @DeityLink and the rest of vg to finish their port of this and make it perform better before having this at all

**Comment by Iamgoofball** [comment]
jesus christ that planet looks fucking horribly out of place

**Comment by PJB3005** [comment]
Yeah it's shit.

**Comment by TheDZD** [comment]
The planet really does look terribly out of place.

**Comment by monster860** [comment]
Do you have something better? I made this in like 5 seconds.

**Comment by TheDZD** [comment]
Not having the planet there, ideally.

**Comment by Aurorablade** [comment]
...no planet at all..

**Comment by NigglyWiggly** [comment]
I hate to drop in on an unrelated codebase like the rest of the /vg/ people but jesus man, that planet looks REALLY out of place.

**Comment by PKPenguin321** [comment]
remove planet

**Comment by DeityLink** [comment]
>I made this in like 5 seconds.

That's your problem here. If you're gonna try to do something, you gotta put all your codermight into it.

**Comment by Fox-McCloud** [comment]
Apparently the Cyberiad now orbits Saturn.

That looks really ick.

**Comment by Spacemanspark** [comment]
Ya, the planet looks really... bleh.
I honestly thought you were joking when you put it there, initially.

**Comment by Twinmold93** [comment]
http://nanotrasen.se/wiki/index.php/Epsilon_Eridani

There is a orange gas giant and a small planet. The question comes about how to... Go about that. At this time, shouldn't have them in the picture.

**Comment by CrAzYPiLoT-SS13** [comment]
Oh God, the planet's hideous.

**Comment by CrAzYPiLoT-SS13** [comment]
I edited some things, and voila!
planet
render_sun

**Comment by Fox-McCloud** [comment]
I'd rather not have a planet at all. :-1:

**Comment by Spacemanspark** [comment]
Ya, let's not have a planet with this, please.

**Comment by PJB3005** [comment]
Maybe a very tiny planet in the background?

**Comment by monster860** [comment]
/vg/ has merged their parallax!

**Comment by monster860** [comment]
conflicts fixed, plz merge.

**Comment by PJB3005** [comment]
Hang on it's not live yet, had we merge it 20 minutes earlier it would've been. The server hasn't updated yet.

**Comment by monster860** [comment]
After seeing the /vg/ test, I definitely think we need to stick with having 1 static layer.

**Comment by tigercat2000** [comment]
>conflicts fixed, plz merge.

/vg/ has a much different version
IMO theirs looks much better

**Comment by PJB3005** [comment]
Ours also killed everybody's FPS for some reason. Even though it worked perfectly fine on the test server.

**Comment by monster860** [comment]
They did a test just now. It lagged horribly, and it wasn't the movement procs. My theory is that there's way too much shit on everyone's screen now. If that's it, then 1 layer should be a lot faster.

**Comment by tigercat2000** [comment]
RIP

**Comment by Spacemanspark** [comment]
I went on to see for myself and I'm getting absolutely no lag at all.

**Comment by DeityLink** [comment]
@monster860 yeah, the lag was caused by screen objects getting toggled, and refreshed. technically, removing the options for players to toggle space dust or toggle parallax (and having 4 objects per layers instead of 9) would fix it. But I'm really tired right now so if we run another test it'll be in a while.

I'm a bit mad that it took 2 weeks for the other guys to finally accept to do a test on the main server, we'd have found those issues back then.

So yeah.

@Spacemanspark You weren't there when the round started. Lots of lag was caused by mob/Login() (and screen objects getting created and added to clients' screens)

**Comment by DeityLink** [comment]
@PJB3005 the version we ran on the test server was vastly different. Also, there were only 3 people on the test server when the round started. Here tonight we had over 60 players become mobs in a single frame. The game froze for almost 10 seconds. And it kept being bad because some players would toggle their parallax nonstop.

**Comment by Spacemanspark** [comment]
Ah, fair enough.

**Comment by PJB3005** [comment]
Hey look, parallax got re-merged on /vg/

https://dl.dropboxusercontent.com/u/107028026/ShareX/2016/06/2016-06-08_17-17-59.webm

IT'S SO PRETTY.

**Comment by Iamgoofball** [comment]
@PJB3005 is it ready for other servers to lift now?

**Comment by PJB3005** [comment]
@Iamgoofball you need to recode it from scratch to use typeless loops first.

**Comment by Iamgoofball** [comment]
welp its dead jim

**Comment by monster860** [comment]
Conflicts fixed.

**Comment by TheDZD** [comment]
Another conflict.

**Comment by Fox-McCloud** [comment]
If/when it doesn't cause copious amounts of lag (which it did, last we heard).

**Comment by monster860** [comment]
Besides the merge conflict, what reason is there to not merge this now? It is much simpler and not prone to the lag that happened on vg

**Comment by monster860** [comment]
Conflicts fixed

**Comment by monster860** [comment]
Conflicts fixed

**Comment by Spacemanspark** [comment]
Please merge this.

**Comment by tigercat2000** [comment]
because /vg/ is actively working on theirs and they actually have more than a single coder to maintain it

**Comment by PJB3005** [comment]
Also the lag was fixed.

**Comment by Spacemanspark** [comment]
yay

**Pull request closed without merging**